File: apps/desktop/src-tauri/src/presence.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum PresenceState {
    Idle,
    Listening,
    Transcribing,
    Routing,
    Planning,
    WaitingForApproval,
    Executing,
    Verifying,
    Recovering,
    Speaking,
    Degraded,
    Failed,
}

impl PresenceState {
    fn label(&self) -> &'static str {
        match self {
            Self::Idle => "Idle",
            Self::Listening => "Listening",
            Self::Transcribing => "Transcribing",
            Self::Routing => "Routing",
            Self::Planning => "Planning",
            Self::WaitingForApproval => "Waiting for approval",
            Self::Executing => "Executing",
            Self::Verifying => "Verifying",
            Self::Recovering => "Recovering",
            Self::Speaking => "Speaking",
            Self::Degraded => "Degraded",
            Self::Failed => "Failed",
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct PresencePayload {
    pub status: PresenceState,
    pub current_task: bool,
    pub pending_approval: bool,
    pub microphone_enabled: bool,
    pub planner_status: String,
    pub stt_status: String,
    pub tts_status: String,
    pub accessibility_status: String,
    pub privacy_mode: String,
}

impl Default for PresencePayload {
    fn default() -> Self {
        Self {
            status: PresenceState::Idle,
            current_task: false,
            pending_approval: false,
            microphone_enabled: true,
            planner_status: "unloaded".into(),
            stt_status: "unloaded".into(),
            tts_status: "unloaded".into(),
            accessibility_status: "not_determined".into(),
            privacy_mode: "ephemeral".into(),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct PresenceLabels {
    pub status: String,
    pub task: String,
    pub approval: String,
    pub microphone: String,
    pub planner: String,
    pub stt: String,
    pub tts: String,
    pub accessibility: String,
    pub privacy: String,
}

impl PresencePayload {
    pub fn labels(&self) -> PresenceLabels {
        PresenceLabels {
            status: format!("Status: {}", self.status.label()),
            task: format!(
                "Current task: {}",
                if self.current_task { "active" } else { "none" }
            ),
            approval: format!(
                "Pending approval: {}",
                if self.pending_approval { "yes" } else { "none" }
            ),
            microphone: format!(
                "Microphone: {}",
                if self.microphone_enabled {
                    "enabled"
                } else {
                    "disabled"
                }
            ),
            planner: format!("Local planner: {}", bounded_status(&self.planner_status)),
            stt: format!("STT: {}", bounded_status(&self.stt_status)),
            tts: format!("TTS: {}", bounded_status(&self.tts_status)),
            accessibility: format!(
                "Accessibility: {}",
                bounded_status(&self.accessibility_status)
            ),
            privacy: format!("Privacy: {}", bounded_status(&self.privacy_mode)),
        }
    }
}

fn bounded_status(value: &str) -> &str {
    const ALLOWED: &[&str] = &[
        "unavailable",
        "unloaded",
        "loading",
        "ready",
        "busy",
        "idle_cached",
        "evicting",
        "degraded",
        "failed",
        "not_determined",
        "denied",
        "granted",
        "revoked",
        "ephemeral",
        "retain_transcripts",
    ];
    if ALLOWED.contains(&value) {
        value
    } else {
        "unavailable"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn presence_payload_labels_are_bounded_and_contain_no_transcript() {
        let payload = PresencePayload {
            status: PresenceState::WaitingForApproval,
            current_task: true,
            pending_approval: true,
            microphone_enabled: true,
            planner_status: "ready".into(),
            stt_status: "degraded".into(),
            tts_status: "ready".into(),
            accessibility_status: "not_determined".into(),
            privacy_mode: "ephemeral".into(),
        };
        let labels = payload.labels();
        assert_eq!(labels.status, "Status: Waiting for approval");
        assert_eq!(labels.task, "Current task: active");
        assert_eq!(labels.approval, "Pending approval: yes");
        let all = format!("{labels:?}").to_lowercase();
        assert!(!all.contains("transcript"));
        assert!(!all.contains("token"));
    }
}

File: apps/desktop/src/lib/presence.ts
import type { TaskState } from "@omnimac/shared-schemas";

export type PresenceStatus =
  | "IDLE"
  | "LISTENING"
  | "TRANSCRIBING"
  | "ROUTING"
  | "PLANNING"
  | "WAITING_FOR_APPROVAL"
  | "EXECUTING"
  | "VERIFYING"
  | "RECOVERING"
  | "SPEAKING"
  | "DEGRADED"
  | "FAILED";

export interface PresenceInput {
  taskState: TaskState | null;
  hasPendingApproval: boolean;
  microphoneEnabled: boolean;
  voiceState: "idle" | "listening" | "transcribing" | "speaking" | "failed";
  plannerState: string;
  sttState: string;
  ttsState: string;
  accessibilityState: string;
  privacyMode: "ephemeral" | "retain_transcripts";
}

export interface PresenceState {
  status: PresenceStatus;
  currentTask: boolean;
  pendingApproval: boolean;
  microphoneEnabled: boolean;
  plannerStatus: string;
  sttStatus: string;
  ttsStatus: string;
  accessibilityStatus: string;
  privacyMode: string;
}

const TERMINAL = new Set<TaskState>([
  "COMPLETED",
  "FAILED",
  "FAILED_REQUIRES_USER",
  "CANCELLED",
]);

export function buildPresenceState(input: PresenceInput): PresenceState {
  let status: PresenceStatus = "IDLE";
  if (input.voiceState === "listening") status = "LISTENING";
  else if (input.voiceState === "transcribing") status = "TRANSCRIBING";
  else if (input.voiceState === "speaking") status = "SPEAKING";
  else if (input.voiceState === "failed") status = "FAILED";
  else if (input.taskState === "UNDERSTANDING") status = "ROUTING";
  else if (input.taskState === "PLANNING") status = "PLANNING";
  else if (input.taskState === "WAITING_FOR_APPROVAL") status = "WAITING_FOR_APPROVAL";
  else if (input.taskState === "EXECUTING") status = "EXECUTING";
  else if (input.taskState === "VERIFYING") status = "VERIFYING";
  else if (input.taskState === "RECOVERING") status = "RECOVERING";
  else if (input.taskState === "FAILED" || input.taskState === "FAILED_REQUIRES_USER") {
    status = "FAILED";
  } else if (input.sttState === "degraded" || input.plannerState === "degraded") {
    status = "DEGRADED";
  }

  return {
    status,
    currentTask: input.taskState !== null && !TERMINAL.has(input.taskState),
    pendingApproval: input.hasPendingApproval,
    microphoneEnabled: input.microphoneEnabled,
    plannerStatus: input.plannerState,
    sttStatus: input.sttState,
    ttsStatus: input.ttsState,
    accessibilityStatus: input.accessibilityState,
    privacyMode: input.privacyMode,
  };
}
